File: python/animals.py
class Animal:
    def __init__(self, name):
        self.name = name
    def speak(self):
        return '{} says {}'.format(self.name, self.sound())

class Cow(Animal):
    def sound(self):
        return 'moooo'

class Horse(Animal):
    def sound(self):
        return 'neigh'

class Sheep(Animal):
    def sound(self):
        return 'baaaa'

if __name__ == '__main__':
    s = Horse('CJ')
    assert s.speak() == 'CJ says neigh'
    c = Cow('Bessie')
    assert c.speak() == 'Bessie says moooo'
    assert Sheep('Little Lamb').speak() == 'Little Lamb says baaaa'

File: go/animals.go
package main

import "fmt"

type Animal struct {
    Sounder
    name string
}

type Sounder interface {
    sound() string
}

func speak(a Animal) string {
    return fmt.Sprintf("%s says %s", a.name, a.sound())
}


type Horse struct {
    Animal
}

func (h Horse) sound() string {
    return "neigh"
}

type Cow struct {
    Animal
}

func (c Cow) sound() string {
    return "moooo"
}

type Sheep struct {
    Animal
}

func (s Sheep) sound() string {
    return "baaaa"
}

func main() {
    h := Animal{Horse{}, "CJ"}
    fmt.Println(speak(h))
    c := Animal{Cow{}, "Bessie"}
    fmt.Println(speak(c))
    fmt.Println(speak(Animal{Sheep{}, "Little Lamb"}))
}
